carmm/analyse/GCN.py

```diff
@@ -0,0 +1,134 @@
+            if i == j:
+                continue
+            elif abs(atoms[i].z-atom_j.z) > 4:
+                continue
+            # Check if atom i and atom j are first nearst neighbours
+            dist = atoms.get_distance(i, j, mic=True)
+            if dist <= bond+0.001 and dist >= bond-0.001:
+                cn += 1
+                fnn.append(j)
+
+        # Append coordination number and first nearest neighbours to the lists every time the second j loop finishes.
+        cn_list.append(cn)
+        fnn_list.append(fnn)
+
+    return cn_list, fnn_list
+
+def fnn_set(fnnLists):
+    """
+    :param fnnLists
+    :return: Set of all fnns
+    """
+    allfnn = []
+    for fnn in fnnLists:
+        allfnn += fnn
+    return set(allfnn)
+    
+def general_coord_number(lattice='fcc', facet=(1,1,1), site='ontop'):
+
+    """
+    GCN calculator.
+    See: 
+    Calle-Vallejo, F. (2023). Advanced Science, 10(20), 2207644. https://doi.org/10.1002/ADVS.202207644
+    Zhao, Z., et al. (2016). Journal of Physical Chemistry C, 120(49), 28125–28130. https://doi.org/10.1021/ACS.JPCC.6B10155/
+    
+    :param lattice: crystal structure
+    :param facet: tuple, e.g. (1,1,1)
+    :param site: 'ontop', 'bridge'
+    :return: gcn. Generalized coordination number
+    """
+    from ase.build import surface, bulk
+    from ase.visualize import view
+    import numpy as np
+
+    sum_fnn_cn = 0 # The sum of CN of fnns
+    a =3.6
+    
+    if lattice == 'fcc':
+        bond = round(a / 2 ** 0.5, 3)
+    elif lattice == 'bcc':
+        bond = round(a * (3 ** 0.5) / 2, 3)
+    
+    if lattice not in ['fcc', 'bcc']:
+        print('Only support fcc and bcc for now')
+        return 'gcn failed'
+    
+
+    copper = bulk('Cu', lattice, a=a, cubic=True)
+    slab = surface(copper, facet, layers=20, vacuum=20)
+     
+    size = len(slab)
+    maxZ = np.max([atom.z for atom in slab])
+
+    lastIndex = 0 # Surface atom at the top right corner
+    toplayerIndices = [atom.index for atom in slab if atom.z == maxZ]
+    toplayerSize = len(toplayerIndices)
+    
+    x=0
+    y=0
+    for index in toplayerIndices:
+        if slab[index].x > x and slab[index].y > y:
+            x = slab[index].x
+            y = slab[index].y
+            lastIndex = index
+
+    
+    # FCC cn_max = 12
+    
+    if site == 'ontop':
+        siteIndices = [lastIndex+5*size,]
+    elif site == 'bridge':
+        # Shift the indices to the centre of the surface
+        topfnn = [atom.index for atom in slab 
+                  if atom.z == maxZ and 
+                  slab.get_distance(atom.index, lastIndex) <= bond+0.001 
+                  and slab.get_distance(atom.index, lastIndex) >= bond-0.001]
+        
+        siteIndices = [lastIndex+5*size, topfnn[0]+5*size]
+
+    slab = slab.repeat((4,4,1))
+
+    print('last index ', lastIndex, 'site indices', siteIndices)
+
+    # Shift the indices to the centre of the surface to bulk interior
```

Comment:
The workflow here is quite involved but I'm not entirely clear what is happening. Can the comments be made a little clearer on the exact need for each step?